# Beginning Work on Twitter Disaster Kaggle Competition

In [1]:
import nltk
import numpy as np
import pandas as pd
import re
from nltk.corpus import stopwords
from sklearn.model_selection import cross_validate, cross_val_score
from sklearn.svm import SVC
from sklearn.linear_model import LogisticRegression
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.neighbors import KNeighborsClassifier
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.tree import DecisionTreeClassifier
from sklearn.naive_bayes import MultinomialNB
import gensim
import gensim.downloader as gensim_api
import warnings
warnings.filterwarnings("ignore")

Below we will define a function that takes in a column with raw text data and returns a cleaned version of that column

In [2]:
train = pd.read_csv("..\\data\\test.csv")
test = pd.read_csv("..\\data\\train.csv")

In [3]:
train.head()

id keyword location                                               text
0   0     NaN      NaN                 Just happened a terrible car crash
1   2     NaN      NaN  Heard about #earthquake is different cities, s...
2   3     NaN      NaN  there is a forest fire at spot pond, geese are...
3   9     NaN      NaN           Apocalypse lighting. #Spokane #wildfires
4  11     NaN      NaN      Typhoon Soudelor kills 28 in China and Taiwan

In [4]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3263 entries, 0 to 3262
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   id        3263 non-null   int64 
 1   keyword   3237 non-null   object
 2   location  2158 non-null   object
 3   text      3263 non-null   object
dtypes: int64(1), object(3)
memory usage: 102.1+ KB


Things to notice is that the size of our dataframe is 3263 but there are only 3237 non-null values and 2158 for keyword and location respectively

In [5]:
train.keyword.unique() # notice NaN is a keyword

array([nan, 'ablaze', 'accident', 'aftershock', 'airplane%20accident',
       'ambulance', 'annihilated', 'annihilation', 'apocalypse',
       'armageddon', 'army', 'arson', 'arsonist', 'attack', 'attacked',
       'avalanche', 'battle', 'bioterror', 'bioterrorism', 'blaze',
       'blazing', 'bleeding', 'blew%20up', 'blight', 'blizzard', 'blood',
       'bloody', 'blown%20up', 'body%20bag', 'body%20bagging',
       'body%20bags', 'bomb', 'bombed', 'bombing', 'bridge%20collapse',
       'buildings%20burning', 'buildings%20on%20fire', 'burned',
       'burning', 'burning%20buildings', 'bush%20fires', 'casualties',
       'casualty', 'catastrophe', 'catastrophic', 'chemical%20emergency',
       'cliff%20fall', 'collapse', 'collapsed', 'collide', 'collided',
       'collision', 'crash', 'crashed', 'crush', 'crushed', 'curfew',
       'cyclone', 'damage', 'danger', 'dead', 'death', 'deaths', 'debris',
       'deluge', 'deluged', 'demolish', 'demolished', 'demolition',
       'derail', 'der

In [6]:
train.location.unique() # again here NaN is a location

array([nan, 'London', "Niall's place | SAF 12 SQUAD |", ...,
       'Acey mountain islanddåÇTorontoåÈ', 'los angeles',
       'Brussels, Belgium'], dtype=object)

Eventually we will need to figure out what to do with these missing values, but for now lets focus on cleaning up the text!

In [7]:
train.id.unique() # no repeating ids which is good

array([    0,     2,     3, ..., 10868, 10874, 10875], dtype=int64)

In [8]:
train.columns

Index(['id', 'keyword', 'location', 'text'], dtype='object')

Found a list of contractions from a stack overflow post. This dictionary will be used to convert contractions to their root words.

In [9]:
# A list of contractions from http://stackoverflow.com/questions/19790188/expanding-english-language-contractions-in-python
contractions = {
"ain't": "am not",
"aren't": "are not",
"can't": "cannot",
"can't've": "cannot have",
"'cause": "because",
"could've": "could have",
"couldn't": "could not",
"couldn't've": "could not have",
"didn't": "did not",
"doesn't": "does not",
"don't": "do not",
"hadn't": "had not",
"hadn't've": "had not have",
"hasn't": "has not",
"haven't": "have not",
"he'd": "he would",
"he'd've": "he would have",
"he'll": "he will",
"he's": "he is",
"how'd": "how did",
"how'll": "how will",
"how's": "how is",
"i'd": "i would",
"i'll": "i will",
"i'm": "i am",
"i've": "i have",
"isn't": "is not",
"it'd": "it would",
"it'll": "it will",
"it's": "it is",
"let's": "let us",
"ma'am": "madam",
"mayn't": "may not",
"might've": "might have",
"mightn't": "might not",
"must've": "must have",
"mustn't": "must not",
"needn't": "need not",
"oughtn't": "ought not",
"shan't": "shall not",
"sha'n't": "shall not",
"she'd": "she would",
"she'll": "she will",
"she's": "she is",
"should've": "should have",
"shouldn't": "should not",
"that'd": "that would",
"that's": "that is",
"there'd": "there had",
"there's": "there is",
"they'd": "they would",
"they'll": "they will",
"they're": "they are",
"they've": "they have",
"wasn't": "was not",
"we'd": "we would",
"we'll": "we will",
"we're": "we are",
"we've": "we have",
"weren't": "were not",
"what'll": "what will",
"what're": "what are",
"what's": "what is",
"what've": "what have",
"where'd": "where did",
"where's": "where is",
"who'll": "who will",
"who's": "who is",
"won't": "will not",
"wouldn't": "would not",
"you'd": "you would",
"you'll": "you will",
"you're": "you are"
}

# This function will be used to do 4 things at once:
 1) it will turn any contractions that appear above into their respected root words i.e "won't" will become "will not"

2) It will use a regular expression to remove any unwanted characters from our text. In this case unwanted characters are punctuation and any special characters sucha as '@'

3) It will remove any stop words that appear in our text. Here the stop words were pulled from a pre-defined list of stopwords from NLTK

4) Lastly it will tokenize our text. Tokenization is the process of taking a body of text and converting it into lists of strings.

In [10]:
def clean_text(text, remove_stopwords = True):
    '''Remove unwanted characters, stopwords, and format the text to create fewer nulls word embeddings'''

    # Convert words to lower case
    text = text.lower()

    # Replace contractions with their longer forms
    if True:
        text = text.split()
        new_text = []
        for word in text:
            if word in contractions:
                new_text.append(contractions[word])
            else:
                new_text.append(word)
        text = " ".join(new_text)

    # Format words and remove unwanted characters
    text = re.sub(r'https?:\/\/.*[\r\n]*', '', text, flags=re.MULTILINE)
    text = re.sub(r'\<a href', ' ', text)
    text = re.sub(r'&amp;', '', text)
    text = re.sub(r'[_"\-;%()|+&=*%.,!?:#$@\[\]/]', ' ', text)
    text = re.sub(r'<br />', ' ', text)
    text = re.sub(r'\'', ' ', text)

    # remove stop words
    if remove_stopwords:
        text = text.split()
        stops = set(stopwords.words("english")) # pulli-ng a list of stopwords from NLTK
        text = [w for w in text if not w in stops]
        text = " ".join(text)

    # Tokenize each word
    text =  nltk.WordPunctTokenizer().tokenize(text)

    return text

In [11]:
train['text_cleaned'] = list(map(clean_text, train.text))
test['text_cleaned'] = list(map(clean_text, test.text))

# Lemmatizing is the process of taking a word like swimming and converting it to its dictionary root word. So swimming becomes swim, eating becomes eat. Swam becomes swim. For more information click here:
https://towardsdatascience.com/lemmatization-in-natural-language-processing-nlp-and-machine-learning-a4416f69a7b6

In [12]:
def lemmatized_words(df):
    lemm = nltk.stem.WordNetLemmatizer()
    df['lemmatized_text'] = list(map(lambda word:
                                     list(map(lemm.lemmatize, word)),
                                     df.text_cleaned))



In [13]:
lemmatized_words(train)
lemmatized_words(test)

In [14]:
train["text_str"]=train["lemmatized_text"].apply(lambda x: " ".join(x) )
test["text_str"]=test["lemmatized_text"].apply(lambda x: " ".join(x) )

In [15]:
train.head()  #ntoice our new dataframe has two new columns text_cleaned and lemmatized_text

id keyword location                                               text  \
0   0     NaN      NaN                 Just happened a terrible car crash   
1   2     NaN      NaN  Heard about #earthquake is different cities, s...   
2   3     NaN      NaN  there is a forest fire at spot pond, geese are...   
3   9     NaN      NaN           Apocalypse lighting. #Spokane #wildfires   
4  11     NaN      NaN      Typhoon Soudelor kills 28 in China and Taiwan   

                                        text_cleaned  \
0                   [happened, terrible, car, crash]   
1  [heard, earthquake, different, cities, stay, s...   
2  [forest, fire, spot, pond, geese, fleeing, acr...   
3         [apocalypse, lighting, spokane, wildfires]   
4      [typhoon, soudelor, kills, 28, china, taiwan]   

                                     lemmatized_text  \
0                   [happened, terrible, car, crash]   
1  [heard, earthquake, different, city, stay, saf...   
2  [forest, fire, spot, pond, goose, fleeing, acr...   
3          [apocalypse, lighting, spokane, wildfire]   
4       [typhoon, soudelor, kill, 28, china, taiwan]   

                                            text_str  
0                        happened terrible car crash  
1  heard earthquake different city stay safe ever...  
2  forest fire spot pond goose fleeing across str...  
3               apocalypse lighting spokane wildfire  
4              typhoon soudelor kill 28 china taiwan

In [16]:
train.loc[train.text.str.find('fuck') > 0] # we will use these explicit words to add an additional column to our dataframe

id            keyword                        location  \
231    750          avalanche                             NaN   
290    937              blaze                   Cleveland, OH   
344   1115          blew%20up                    Coos Bay, OR   
353   1140             blight                             NaN   
377   1210           blizzard                      That place   
411   1330         blown%20up                     Scout Team    
414   1339         blown%20up              somewhere or other   
481   1564               bomb                        shanghai   
584   1902            burning                  daily ? 18 ? ?   
666   2168       catastrophic                             NaN   
704   2291       cliff%20fall                             NaN   
826   2716            crushed                      Motown, WV   
846   2779             curfew        770 to Benedict College    
878   2897             damage                 California, USA   
1018  3348         demolished                     Connecticut   
1020  3350         demolished                             NaN   
1432  4743           evacuate                       San Diego   
1451  4815         evacuation                             NaN   
1482  4932           exploded                             NaN   
1576  5318               fire                             NaN   
1629  5483             flames                     Little Rock   
1804  6095           hellfire                             NaN   
2190  7333  nuclear%20reactor                             NaN   
2552  8521          screaming         HYPE RESSHA HYPE RESSHA   
2956  9793            trapped                     Houston, TX   
2961  9804            trapped                         Orlando   
3018  9966            tsunami  rest easy angel @datboiiharri.   

                                                   text  \
231   if this fucking is true i will be decapitated ...   
290   I never got paid to give a fuck..we might as w...   
344   We were fucking around on Google maps at work ...   
353                 @colemcfadyean fuck off cole ??????   
377   If blizzard did another 12 month sub thing and...   
411   If you bored as shit don't nobody fuck wit you...   
414   I don't understand how White Iverson by Post M...   
481      Bruh this sex on the beach??is bomb as fuck yo   
584   RT: A real burn book entry about CA: why the f...   
666   If a å£1 rise in wages is going to have such a...   
704   @JohnnieGuilbert jump off a cliff. I'll be che...   
826   So in one episode they undo season 1. Kai join...   
846   everybody like fuck curfew.. and i'm sitting h...   
878                Got my first damage today  fuckkkkkk   
1018                   That ball was fucking demolished   
1020  @who_mikejoness I hate seeing the boy get fuck...   
1432  my father fucking died when the north tower co...   
1451  my school is so fucking dumb they just set off...   
1482  Our electricity just fucking exploded they're ...   
1576  almost set the house on fire tryna make fuckin...   
1629           I only fuck you when it's half past five   
1804  I'm melting a bar of chocolate under my laptop...   
2190  they're gonna fucking steal a nuclear fission ...   
2552       i'm fucking screaming http://t.co/H3MrqRrGfe   
2956  If I ever saw a dog trapped in a hot car you b...   
2961  It's a trap not a fucking game. U ain't trappi...   
3018  I just wanna food but of course I had to get c...   

                                           text_cleaned  \
231   [fucking, true, decapitated, throw, head, aval...   
290   [never, got, paid, give, fuck, might, well, bl...   
344   [fucking, around, google, maps, work, pulled, ...   
353                         [colemcfadyean, fuck, cole]   
377   [blizzard, another, 12, month, sub, thing, gav...   
411   [bored, shit, nobody, fuck, wit, busy, yo, shi...   
414   [understand, white, iverson, post, malone, blo...   
481                  [bruh, sex, beach, bomb, fuck, yo]   
584

## That was some basic cleaning, lemmatizing, and tokenizing: Now let us try to prepare our model

In [19]:
# Create a Vectorizer Object
vectorizer = CountVectorizer()

vectorizer.fit(test.text_str)

# Printing the identified Unique words along with their indices
print("Vocabulary: ", vectorizer.vocabulary_)

# Encode the Document
X_varied = vectorizer.transform(test.text_str)

# Summarizing the Encoded Texts
print("Encoded Document is:")
print(X_varied.toarray())

Vocabulary:  {'deed': 3757, 'reason': 10932, 'earthquake': 4433, 'may': 8479, 'allah': 916, 'forgive': 5377, 'forest': 5369, 'fire': 5210, 'near': 9203, 'la': 7690, 'ronge': 11415, 'sask': 11659, 'canada': 2467, 'resident': 11171, 'asked': 1296, 'shelter': 12000, 'place': 10233, 'notified': 9426, 'officer': 9572, 'evacuation': 4823, 'order': 9738, 'expected': 4909, '13': 84, '000': 1, 'people': 10061, 'receive': 10951, 'wildfire': 14616, 'california': 2429, 'got': 5861, 'sent': 11868, 'photo': 10147, 'ruby': 11487, 'alaska': 873, 'smoke': 12294, 'pours': 10396, 'school': 11727, 'rockyfire': 11377, 'update': 14045, 'hwy': 6602, '20': 181, 'closed': 2961, 'direction': 4032, 'due': 4362, 'lake': 7714, 'county': 3339, 'cafire': 2402, 'flood': 5283, 'disaster': 4054, 'heavy': 6250, 'rain': 10818, 'cause': 2615, 'flash': 5254, 'flooding': 5285, 'street': 12772, 'manitou': 8341, 'colorado': 3068, 'spring': 12558, 'area': 1195, 'top': 13578, 'hill': 6348, 'see': 11815, 'wood': 14727, 'emergenc

In [20]:
y_varied = test[['target']]

# logistic regression
logit_clf = LogisticRegression()
logit_clf.fit(X_varied, y_varied)
y_logit = logit_clf.predict(X_varied)

# svm
svm_clf = SVC()
svm_clf.fit(X_varied, y_varied)
y_svm = svm_clf.predict(X_varied)


# scoring (naive, not cross-validated)
print(f"Logit score: {logit_clf.score(X_varied, y_varied):.3f}")
print(f"SVM score:   {svm_clf.score(X_varied, y_varied):.3f}")

# scoring (with cross validation)
print(f"Logit score, cross validated: {np.mean(cross_val_score(logit_clf, X_varied, y_varied)):.3f}")
print(f"SVM score, cross validated:   {np.mean(cross_val_score(svm_clf, X_varied, y_varied)):.3f}")

Logit score: 0.948
SVM score:   0.942
Logit score, cross validated: 0.693
SVM score, cross validated:   0.706


In [21]:
# for n in range(1,10):
#     #knn
#     print(f"Number of clusters is {n}")
#     knn_clf = KNeighborsClassifier(n_neighbors=n) # instantiate
#     knn_clf.fit(X_varied, y_varied)               # fit
#     y_knn = knn_clf.predict(X_varied)             # predict
#
#     #scoring (naive, not cross-validated)
#     print(f"kNN score:   {knn_clf.score(X_varied, y_varied):.3f}")
#
#     #scoring (with cross validation)
#     print(f"kNN score, cross validated:   {np.mean(cross_val_score(knn_clf, X_varied, y_varied)):.3f}\n")

# Now that we have used a count vectorizer to transformn our cornpus let us try to use a TF-ID

In [22]:
vectorizer = TfidfVectorizer()

X_varied = vectorizer.fit_transform(test.text_str)

vectorizer.get_feature_names_out()


print(X_varied.shape)

(7613, 15082)


In [23]:
y_varied = test[['target']]

# logistic regression
logit_clf = LogisticRegression()
logit_clf.fit(X_varied, y_varied)
y_logit = logit_clf.predict(X_varied)

# svm
svm_clf = SVC()
svm_clf.fit(X_varied, y_varied)
y_svm = svm_clf.predict(X_varied)

# scoring (naive, not cross-validated)
print(f"Logit score: {logit_clf.score(X_varied, y_varied):.3f}")
print(f"SVM score:   {svm_clf.score(X_varied, y_varied):.3f}")

# scoring (with cross validation)
print(f"Logit score, cross validated: {np.mean(cross_val_score(logit_clf, X_varied, y_varied)):.3f}")
print(f"SVM score, cross validated:   {np.mean(cross_val_score(svm_clf, X_varied, y_varied)):.3f}\n")



Logit score: 0.883
SVM score:   0.962
Logit score, cross validated: 0.705
SVM score, cross validated:   0.698



In [24]:
# for n in range(7, 15): # it seens like 11 is the best score for our model
#     #knn
#     print(f"Number of clusters is {n}")
#     knn_clf = KNeighborsClassifier(n_neighbors=n)  # instantiate
#     knn_clf.fit(X_varied, y_varied)  # fit
#     y_knn = knn_clf.predict(X_varied)  # predict
#
#     #scoring (naive, not cross-validated)
#     print(f"kNN score:   {knn_clf.score(X_varied, y_varied):.3f}")
#
#     #scoring (with cross validation)
#     print(f"kNN score, cross validated:   {np.mean(cross_val_score(knn_clf, X_varied, y_varied)):.3f}\n")

In [25]:
# Classifers to test
classifiers = {
    'kNN': KNeighborsClassifier(n_neighbors = 11),
    'Logit':LogisticRegression(),
    'Tree': DecisionTreeClassifier(),
    'Bayes': MultinomialNB()
}

scores1 = {} # Store cross-validation results in a dictionary
for classifier in classifiers:
    scores1[classifier] = cross_validate( # perform cross-validation
        classifiers[classifier], # classifier object
        X_varied, # feature matrix
        y_varied, # gold labels
        cv=10, #number of folds
        scoring=['accuracy', 'recall', 'f1'] # scoring methods
    )

# Lets try the same approach but without removing stopwords

In [26]:
train['text_lower'] = train.text.apply(lambda x: x.lower())
test['text_lower'] = test.text.apply(lambda x: x.lower())
test.columns

Index(['id', 'keyword', 'location', 'text', 'target', 'text_cleaned',
       'lemmatized_text', 'text_str', 'text_lower'],
      dtype='object')

In [27]:
vectorizer.fit(test['text_lower'])
X = vectorizer.transform(test['text_lower'])
y =  test[["target"]]

# Classifers to test
classifiers = {
    'kNN': KNeighborsClassifier(n_neighbors = 11),
    'Logit':LogisticRegression(),
    'Tree': DecisionTreeClassifier(),
    'Bayes': MultinomialNB()
}

scores2 = {} # Store cross-validation results in a dictionary
for classifier in classifiers:
    scores2[classifier] = cross_validate( # perform cross-validation
        classifiers[classifier], # classifier object
        X, # feature matrix
        y, # gold labels
        cv=10, #number of folds
        scoring=['accuracy', 'recall', 'f1'] # scoring methods
    )

In [28]:
# Examine the performance of our simple classifiers

def compare_scores(scores_dict):
    '''
    Takes a dictionary of cross_validate scores.
    Returns a color-coded Pandas dataframe that summarizes those scores.
    '''
    df = pd.DataFrame(scores_dict).T.applymap(np.mean).style.background_gradient(cmap='RdYlGn')
    return df
# Compare cross-validation scores

In [29]:
compare_scores(scores1)

In [30]:
compare_scores(scores2)

# Time for a different apporach. Word Embeddings

In [31]:
from transformers import AutoTokenizer, AutoModelForSeq2SeqLM

tokenizer = AutoTokenizer.from_pretrained("mrm8488/t5-base-finetuned-sarcasm-twitter")

model = AutoModelForSeq2SeqLM.from_pretrained("mrm8488/t5-base-finetuned-sarcasm-twitter")

Downloading:   0%|          | 0.00/850M [00:00<?, ?B/s]

In [34]:
def eval_conversation(text):

    input_ids = tokenizer.encode(text + '</s>', return_tensors='pt')

    output = model.generate(input_ids=input_ids, max_length=3)

    dec = [tokenizer.decode(ids) for ids in output]

    label = dec[0]

    return label

# We will get 'normal' when sarcasm is not detected and 'derison' when detected


In [35]:
train["sarcasm"] =train.text_lower.apply(eval_conversation)

In [40]:
#train.to_csv("train_cleaned.csv")

In [41]:
train

id keyword location  \
0         0     NaN      NaN   
1         2     NaN      NaN   
2         3     NaN      NaN   
3         9     NaN      NaN   
4        11     NaN      NaN   
...     ...     ...      ...   
3258  10861     NaN      NaN   
3259  10865     NaN      NaN   
3260  10868     NaN      NaN   
3261  10874     NaN      NaN   
3262  10875     NaN      NaN   

                                                   text  \
0                    Just happened a terrible car crash   
1     Heard about #earthquake is different cities, s...   
2     there is a forest fire at spot pond, geese are...   
3              Apocalypse lighting. #Spokane #wildfires   
4         Typhoon Soudelor kills 28 in China and Taiwan   
...                                                 ...   
3258  EARTHQUAKE SAFETY LOS ANGELES ‰ÛÒ SAFETY FASTE...   
3259  Storm in RI worse than last hurricane. My city...   
3260  Green Line derailment in Chicago http://t.co/U...   
3261  MEG issues Hazardous Weather Outlook (HWO) htt...   
3262  #CityofCalgary has activated its Municipal Eme...   

                                           text_cleaned  \
0                      [happened, terrible, car, crash]   
1     [heard, earthquake, different, cities, stay, s...   
2     [forest, fire, spot, pond, geese, fleeing, acr...   
3            [apocalypse, lighting, spokane, wildfires]   
4         [typhoon, soudelor, kills, 28, china, taiwan]   
...                                                 ...   
3258  [earthquake, safety, los, angeles, ‰, ûò, safe...   
3259  [storm, ri, worse, last, hurricane, city3other...   
3260                 [green, line, derailment, chicago]   
3261    [meg, issues, hazardous, weather, outlook, hwo]   
3262  [cityofcalgary, activated, municipal, emergenc...   

                                        lemmatized_text  \
0                      [happened, terrible, car, crash]   
1     [heard, earthquake, different, city, stay, saf...   
2     [forest, fire, spot, pond, goose, fleeing, acr...   
3             [apocalypse, lighting, spokane, wildfire]   
4          [typhoon, soudelor, kill, 28, china, taiwan]   
...                                                 ...   
3258  [earthquake, safety, los, angeles, ‰, ûò, safe...   
3259  [storm, ri, worse, last, hurricane, city3other...   
3260                 [green, line, derailment, chicago]   
3261     [meg, issue, hazardous, weather, outlook, hwo]   
3262  [cityofcalgary, activated, municipal, emergenc...   

                                               text_str  \
0                           happened terrible car crash   
1     heard earthquake different city stay safe ever...   
2     forest fire spot pond goose fleeing across str...   
3                  apocalypse lighting spokane wildfire   
4                 typhoon soudelor kill 28 china taiwan   
...                                                 ...   
3258  earthquake safety los angeles ‰ ûò safety fast...   
3259  storm ri worse last hurricane city3others hard...   
3260                      green line derailment chicago   
3261            meg issue hazardous weather outlook hwo   
3262  cityofcalgary activated municipal emergency pl...   

                                             text_lower           sarcasm  
0                    just happened a terrible car crash     <pad> derison  
1     heard about #earthquake is different cities, s...  <pad> normal</s>  
2     there is a forest fire at spot pond, geese are...  <pad> normal</s>  
3              apocalypse lighting. #spokane #wildfires     <pad> derison  
4         typhoon soudelor kills 28 in china and taiwan     <pad> derison  
...                                                 ...               ...  
3258  earthquake safety los angeles ‰ûò safety faste...  <pad> normal</s>  
3259  storm in ri worse than last hurricane. my city...  <pad> normal</s>  
3260  green line derailment in chicago http://t.co/u...     <pad> derison  
3261  meg issues hazardous w